File: TasteKit/tastekit/packages/core/schemas/trace.ts
import { z } from 'zod';

/**
 * Trace Event v1 Schema
 * 
 * Machine-readable trace events for all operations.
 * Traces are written as JSONL (one event per line).
 */

export const TraceEventSchema = z.object({
  schema_version: z.literal('trace_event.v1'),
  run_id: z.string().describe('Unique run identifier'),
  timestamp: z.string().datetime().describe('ISO8601 timestamp'),
  
  actor: z.enum(['agent', 'user', 'system']),
  skill_id: z.string().optional(),
  playbook_id: z.string().optional(),
  step_id: z.string().optional(),
  
  event_type: z.enum([
    'plan',
    'think',
    'tool_call',
    'tool_result',
    'approval_requested',
    'approval_response',
    'artifact_written',
    'memory_write',
    'evaluation',
    'error',
  ]),
  
  tool_ref: z.string().optional().describe('MCP server/tool reference'),
  input_hash: z.string().optional().describe('Hash of input data'),
  output_hash: z.string().optional().describe('Hash of output data'),
  risk_score: z.number().min(0).max(1).optional(),
  principle_refs: z.array(z.string()).optional().describe('Constitution principle IDs'),
  cost: z.object({
    tokens: z.number().optional(),
    time_ms: z.number().optional(),
  }).optional(),
  
  data: z.record(z.any()).optional().describe('Event-specific data'),
  error: z.string().optional().describe('Error message if event_type is error'),
});

export type TraceEvent = z.infer<typeof TraceEventSchema>;

File: AutoClaw/internal/artifact/trace.go
package artifact

// TraceEvent represents a single event in a JSONL trace log.
// Loaded from .tastekit/traces/*.jsonl (one JSON object per line).
type TraceEvent struct {
	SchemaVersion string `json:"schema_version" yaml:"schema_version"`
	RunID         string `json:"run_id" yaml:"run_id"`
	Timestamp     string `json:"timestamp" yaml:"timestamp"` // ISO8601

	Actor      string  `json:"actor" yaml:"actor"`                                 // "agent"|"user"|"system"
	SkillID    *string `json:"skill_id,omitempty" yaml:"skill_id,omitempty"`
	PlaybookID *string `json:"playbook_id,omitempty" yaml:"playbook_id,omitempty"`
	StepID     *string `json:"step_id,omitempty" yaml:"step_id,omitempty"`

	// Event classification. One of: plan, think, tool_call, tool_result,
	// approval_requested, approval_response, artifact_written, memory_write,
	// evaluation, error.
	EventType string `json:"event_type" yaml:"event_type"`

	ToolRef       *string  `json:"tool_ref,omitempty" yaml:"tool_ref,omitempty"`
	InputHash     *string  `json:"input_hash,omitempty" yaml:"input_hash,omitempty"`
	OutputHash    *string  `json:"output_hash,omitempty" yaml:"output_hash,omitempty"`
	RiskScore     *float64 `json:"risk_score,omitempty" yaml:"risk_score,omitempty"` // 0.0-1.0
	PrincipleRefs []string `json:"principle_refs,omitempty" yaml:"principle_refs,omitempty"`

	Cost  *TraceEventCost    `json:"cost,omitempty" yaml:"cost,omitempty"`
	Data  map[string]any     `json:"data,omitempty" yaml:"data,omitempty"`
	Error *string            `json:"error,omitempty" yaml:"error,omitempty"`
}

// TraceEventCost captures resource usage for a single trace event.
type TraceEventCost struct {
	Tokens *int `json:"tokens,omitempty" yaml:"tokens,omitempty"`
	TimeMs *int `json:"time_ms,omitempty" yaml:"time_ms,omitempty"`
}
